---
title: "SC3 package manual"
author: "Vladimir Kiselev"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
vignette: >
  %\VignetteIndexEntry{SC3 package manual}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png')
```

# Introduction

Single-Cell Consensus Clustering (__SC3__) is a tool for unsupervised clustering of scRNA-seq data. __SC3__ achieves high accuracy and robustness by consistently integrating different clustering solutions through a consensus approach. An interactive graphical implementation makes __SC3__ accessible to a wide audience of users. In addition, __SC3__ also aids biological interpretation by identifying marker genes, differentially expressed genes and outlier cells. A manuscript describing __SC3__ in details is currently under review but a copy of it is available on [bioRxiv](http://biorxiv.org/content/early/2016/09/02/036558).

# `scater` and QC

__SC3__ does not perform any type of the sequencing quality control (QC). Therefore we recommend the users to perform their own QC analysis. To encourage the QC, __SC3__ is built on top of the Bioconductorâ€™s `scater` package. The `scater` package contains tools to help with the analysis and QC of single-cell RNA-Seq data.

The basic data container that is used in `scater` is an `SCESet` object. __SC3__ implements several methods that allow one to perform clustering of the expression data contained in the `SCESet` object.

# Quick Start

## SC3 Input

If you already have an `SCESet` object created and QCed using `scater` then proceed to the next chapter.

If you have a matrix containing expression data that was QCed by some other tool, then we first need to form an `SCESet` object containing the data. For this we will use an example of expression matrix provided with `SC3`. The dataset represent gene expression of 80 cells derived from the distal lung epithelium of mice. The authors ([Treutlein et al.](http://www.nature.com/nature/journal/v509/n7500/full/nature13173.htm)) had computationally identified 5 clusters in this dataset. The rows in the `treutlein` dataset correspond to genes and columns correspond to cells. Column names correspond to clusters identified by the authors.

```{r, message=FALSE, warning=FALSE}
library(scater)
library(SC3)
treutlein[1:3, 1:3]
```

Let us prepare our dataset for `SCESet` object creation. Here we follow the [`scater` manual](https://www.bioconductor.org/packages/release/bioc/vignettes/scater/inst/doc/vignette.html):

```{r}
sc_example_cell_info <- data.frame(Author_Cluster = colnames(treutlein))
sc_example_counts <- treutlein
cell_inds <- paste("Cell", 1:80, sep = "_")
rownames(sc_example_cell_info) <- cell_inds
colnames(sc_example_counts) <- cell_inds
```

```{r}
pd <- new("AnnotatedDataFrame", data = sc_example_cell_info)
example_sceset <- newSCESet(countData = sc_example_counts, phenoData = pd)
```

Note that `sc_example_cell_info` data frame can contain more information about the cells, such as plate, run, date etc.

## `@consensus` slot

## Run SC3

```{r, include=FALSE}
# this is only for this vignette - otherwise will not compile
# example_sceset@consensus$sc3_n_cores <- 1
```

```{r, message=FALSE, warning=FALSE}
example_sceset <- sc3(example_sceset, ks = 3:5)
```

You can then visualise the results using an interactive Shiny application:
```{r, eval=FALSE}
sc3_interactive(example_sceset)
```

Visual exploration can provide a reasonable estimate of the number of clusters k. You can then summarise all results into a single list:
```{r}
example_sceset <- sc3_summarise_results(example_sceset, k = 3)
```

This will write to .... You can also export the results into Excel file:
```{r eval=FALSE}
library(WriteXLS)
WriteXLS(
    example_sceset@consensus$sc3_results, 
    ExcelFileName = "sc3_results.xls",
    SheetNames = names(example_sceset@consensus$sc3_results),
    row.names = TRUE
)
```

# Plot Functions

SC3 provides methods for plotting all figures from the interactive session.

## Consensus Matrix

The consensus matrix is a NxN matrix, where N is the number of cells. It represents similarity between the cells based on the averaging of clustering results from all  combinations of clustering parameters. Similarity 0 (blue) means that the two cells are always assigned to different clusters. In contrast, similarity 1 (red) means that the two cells are always assigned to the same cluster. The consensus matrix is clustered by hierarchical clustering and has a diagonal-block structure. Intuitively, the perfect  clustering is achieved when all diagonal blocks are completely red and all off-diagonal elements are completely blue.
```{r}
sc3_plot_consensus(example_sceset, k = 3)
```

## Silhouette Plot

A silhouette is a quantitative measure of the diagonality of the consensus matrix. An average silhouette width (shown at the bottom left of the silhouette plot) varies from 0 to 1, where 1 represents a perfectly block-diagonal consensus matrix and 0 represents a situation where there is no block-diagonal structure. The best clustering is achieved when the average silhouette width is close to 1.
```{r}
sc3_plot_silhouette(example_sceset, k = 3)
```

## Expression Matrix

The expression panel represents the original input expression matrix (cells in columns and genes in rows) after cell and gene filters. Genes are clustered by kmeans with k = 100 (dendrogram on the left) and the heatmap represents the expression levels of the gene cluster centers after log2-scaling.
```{r}
sc3_plot_expression(example_sceset, k = 3)
```

## Cluster Stability

Stability index shows how stable each cluster is accross the selected range of ks. The stability index varies between 0 and 1, where 1 means that the same cluster appears in every solution for different k.
```{r}
sc3_plot_cluster_stability(example_sceset, k = 3)
```

## tSNE plot

[tSNE](https://lvdmaaten.github.io/tsne/) (t-Distributed Stochastic Neighbor Embedding) method is used to map high-dimensional data to a 2D space while preserving local distances between cells. tSNE has become a very popular visualisation tool. SC3 imports the Rtsne function from the [Rtsne package](https://cran.r-project.org/web/packages/Rtsne/index.html) to perform the tSNE analysis. The colors on the plot correspond to the clusters identified by SC3. One of the most sensitive parameters in tSNE analysis is the so-called perplexity. SC3 defines the default perplexity as N/5, where N is the number of cells.
```{r}
sc3_plot_tsne(example_sceset, k = 3)
```

## DE genes

Differential expression is calculated using the non-parametric Kruskal-Wallis test. A significant p-value indicates that gene expression in at least one cluster stochastically dominates one other cluster. SC3 provides a list of all differentially expressed genes with adjusted p-values < 0.01 and plots gene expression profiles of the 50 genes with the lowest p-values. Note that the calculation of differential expression after clustering can introduce a bias in the distribution of p-values, and thus we advise to use the p-values for ranking the genes only.
```{r, fig.height=8}
sc3_plot_de_genes(example_sceset, k = 3)
```

## Marker Genes

To find marker genes, for each gene a binary classifier is constructed based on the mean cluster expression values. The classifier prediction is then calculated using the gene expression ranks. The area under the receiver operating characteristic (ROC) curve is used to quantify the accuracy of the prediction. A p-value is assigned to each gene by using the Wilcoxon signed rank test. By default the genes with the area under the ROC curve (AUROC) > 0.85 and with the p-value < 0.01 are selected and the top 10 marker genes of each cluster are visualized in this heatmap.
```{r, fig.height=6}
sc3_plot_markers(example_sceset, k = 3)
```

## Cell Outliers

Outlier cells in each cluster are detected using robust distances, calculated using the minimum covariance determinant (MCD). The outlier score shows how different a cell is from all other cells in the cluster and it is defined as the differences between the square root of the robust distance and the square root of the 99.99% quantile of the Chi-squared distribution.
```{r}
sc3_plot_cell_outliers(example_sceset, k = 3)
```


```{r, include=FALSE, eval=FALSE}
example_sceset <- sc3_prepare(example_sceset)
example_sceset <- sc3_estimate_k(example_sceset)
example_sceset <- sc3_calc_dists(example_sceset)
example_sceset <- sc3_calc_transfs(example_sceset)
example_sceset <- sc3_kmeans(example_sceset)
example_sceset <- sc3_calc_consens(example_sceset)
example_sceset <- sc3_calc_biology(example_sceset)
# sc3_interactive(example_sceset)
example_sceset <- sc3_summarise_results(example_sceset, k = 3)
```


```{r, include=FALSE, eval=FALSE}
example_sceset@consensus <- list()
example_sceset <- sc3_prepare(example_sceset, svm.num.cells = 50)
example_sceset <- sc3_estimate_k(example_sceset)
# this is only for this vignette - otherwise will not compile
# example_sceset@consensus$sc3_n_cores <- 1
example_sceset <- sc3_calc_dists(example_sceset)
example_sceset <- sc3_calc_transfs(example_sceset)
example_sceset <- sc3_kmeans(example_sceset, ks = 3:5)
example_sceset <- sc3_calc_consens(example_sceset)
example_sceset <- sc3_calc_biology(example_sceset)
# sc3_interactive(example_sceset)
example_sceset <- sc3_run_svm(example_sceset, k = 3)
example_sceset <- sc3_summarise_results(example_sceset, k = 3)
```


# __SC3__ parameters

By default __SC3__ clusters the input dataset in a region of __k__ from 3 to 7. This region can be changed by adjusting the __ks__ parameter of __SC3__. E.g. if you would like to check the clustering of your own __dataset__ for __k__ from 2 to 5, then you need to run the following:
```{r, eval=FALSE}
sc3(dataset, ks = 2:5)
```

where __dataset__ is either an R matrix / data.frame / data.table object OR a path to your input file containing an expression matrix.

For the full list of __SC3__ parameters please read the documentation by typing ?sc3.

# Number of cells

Combining multiple clustering results through the consensus approach is quite computationally expensive. Therefore running __SC3__ on datasets with 1,000 cells may take ~10-15 mins. We do not recommend to use the default options on datasets with more than 1,000 cells. In order to  apply  SC3 to larger datasets, we have implemented a hybrid approach that combines unsupervised and supervised methodologies. When the number of cells is large (N>1,000), SC3 selects a small subset of cells uniformly at random, and obtains clusters from this subset. Subsequently, the inferred labels are used to train a Support Vector Machine (SVM), which is employed to assign labels to the remaining cells. Training the SVM typically takes only a few minutes, thus allowing SC3 to be applied to very large datasets. Based on our results, we set the default of __SC3__ so that when N>1,000 only 20% of the cells are used and when N>5,000 only 1,000 cells are used for clustering before training the SVM. To enable the SVM training please set the _svm_ parameter to __TRUE__:

```{r, eval=FALSE}
sc3(treutlein, svm = TRUE)
```

You can also control the number of training cells used to train SVM with the _svm.num.cells_ parameter:

```{r, eval=FALSE}
sc3(treutlein, svm = TRUE, svm.num.cells = 25)
```

# Input file format

To run __SC3__ on an input file containing an expression matrix one need to preprocess the input file so that it looks as follows:


|  | cell1 | cell2 | cell3 | cell4 | cell5
--- | --- | --- | --- | --- | ---
| __gene1__ | 1 | 2 | 3 | 4 | 5
| __gene2__ | 1 | 2 | 3 | 4 | 5
| __gene3__ | 1 | 2 | 3 | 4 | 5


The first row of the expression matrix (with cell labels, e.g. __cell1__, __cell2__, etc.) should contain one fewer field than all other rows. Separators should be either spaces or tabs. If separators are commas (,) then the extension of the file must be .csv. If a path to your input file is "/path/to/input/file/expression-matrix.txt", to run it:
```{r, eval=FALSE}
sc3("/path/to/input/file/expression-matrix.txt", ks = 2:5)
```

# Saving results

After finding the best clustering solution __SC3__ allows a user to either export the results into an excel file or to save them to the current R session. If the former is chosen then the excel file will be written to the default _Downloads_ folder used by your browser. When saving to the R session a variable (list) __SC3.results__ containing all the results from the interactive session will be created.

# Prerequisites (optional)

> Note that this prerequisite is only optional. If it is not installed you can still run __SC3__, but the _GO anlaysis_ functionality will be disabled.

__SC3__ imports some of the [RSelenium](https://cran.r-project.org/web/packages/RSelenium/) functionality. [RSelenium](https://cran.r-project.org/web/packages/RSelenium/) depends on a stand-alone java binary file (see [Rselenium documentation](https://cran.r-project.org/web/packages/RSelenium/vignettes/RSelenium-basics.html) for more details). You can download and install this binary file by running (the file size is about 30Mb):

```{r, eval=FALSE}
RSelenium::checkForServer()
```

Note, this command has to be executed only once, before running __SC3__ for the first time. Also note that the minimum Java version requirement for [RSelenium](https://cran.r-project.org/web/packages/RSelenium/) is 1.7 (see [this post](https://github.com/ropensci/RSelenium/issues/54) for details).
